### DISCUSSION TITLE
Inferring type from unknown zod schema

### DISCUSSION THREAD

**Discussion by faustbrian** [colinhacks/zod#2301] [Q&A]
So what I'm trying to achieve is that I have a base class where a zod schema gets specified in a `getArguments` function and that schema then gets used by a `parse` function to parse and validate an object. The value that `parse` returns should be typed so that editors can provide autocompletion but at the moment the return type is `{ [x: string]: any }` due to `z.infer` not knowing the exact structure ahead of time. A bit of code to illustrate things:

```ts
import { z } from "zod";

export abstract class AbstractCommand {
  protected parse(args) {
    const schema = this.getArguments();

    type Result = z.infer<typeof schema>;

    return schema.parse(args) as Result; // Also tried `{ [Key in keyof Result]: Result[Key] }`
  }

  protected abstract getArguments();
}

class HelpCommand extends AbstractCommand {
  public handle() {
    // parse() would ideally now have autocompletion based on its zod schema type
    this.parse({ username: 'johndoe' }).username;
  }

  protected override getArguments() {
    return z.object({
      username: z.string(),
    });
  }
}
```

Would there be any way to have `z.infer` infer the the type of the schema that gets returned by `getArguments` without knowing the exact return type of `getArguments` other than it being a zod schema?

**Comment by yosr-fourati** [comment]
Unfortunately, it is not possible to infer the type of the schema returned by the getArguments method without knowing the exact return type of getArguments. This is because TypeScript's type inference is based on the static type information available at compile-time.

In your case, the AbstractCommand class is abstract and does not specify the return type of the getArguments method. Therefore, TypeScript cannot infer the exact schema type that will be returned by getArguments. As a result, the Result type in the parse method will always be inferred as { [x: string]: any }.

To solve this issue, you can explicitly specify the schema type in the AbstractCommand class and use it as the return type of the getArguments method. For example:
```
import { z, ZodType } from "zod";

type SchemaType<T extends ZodType<any, any>> = T extends ZodType<infer U, any> ? U : never;

export abstract class AbstractCommand<T extends ZodType<any, any>> {
  protected abstract getArguments(): T;

  protected parse(args: any): SchemaType<T> {
    const schema = this.getArguments();
    return schema.parse(args);
  }
}

class HelpCommand extends AbstractCommand<typeof z.object({
  username: z.string(),
})> {
  public handle() {
    const parsedArgs = this.parse({ username: 'johndoe' });
    console.log(parsedArgs.username); // "johndoe"
  }

  protected override getArguments() {
    return z.object({
      username: z.string(),
    });
  }
}
```
In this example, the AbstractCommand class is generic and takes a T-type parameter that extends ZodType<any, any>. The getArguments method returns a value of type T. The parse method uses the T type parameter to infer the return type of the schema. parse call. The SchemaType type helper is used to extract the inferred type of the schema.

The HelpCommand class extends AbstractCommand and specifies the exact schema type that it expects to receive from the getArguments method. This allows TypeScript to infer the exact schema type and provide autocompletion for the parsed object.
